### PR TITLE
Changed is_async_connected to use epoll when available

### PR DESCRIPTION
**Pull request by astere-cpr**
Scalability improvement to remove the 1024 socket limit imposed by select() on Linux

### CODE REVIEW

File: clients/roscpp/src/libros/io.cpp [hunk 1]
```diff
@@ -380,6 +380,26 @@ pollfd_vector_ptr poll_sockets(int epfd, socket_pollfd *fds, nfds_t nfds, int ti
...
+    if (fd_cnt < 0) {
+      if (errno != EINTR) {
+        ROS_ERROR("Error in epoll_wait! %s", strerror(errno));
+        err = errno;
+        return -1;
+      }
+    } else if (fd_cnt == 0) {
+      return 0;
+    } else if (ev[0].events & EPOLLERR || ev[0].events & EPOLLHUP) {
+      return -1;
```
Comment (dirk-thomas) [inline clients/roscpp/src/libros/io.cpp:405]: Shouldn't `err` be set to something here?

File: clients/roscpp/src/libros/io.cpp [hunk 2]
```diff
@@ -380,6 +380,26 @@ pollfd_vector_ptr poll_sockets(int epfd, socket_pollfd *fds, nfds_t nfds, int ti
...
+    int fd_cnt = ::epoll_wait(epfd, ev, nfds, 0);
+    close_socket_watcher(epfd);
+    if (fd_cnt < 0) {
+      if (errno != EINTR) {
+        ROS_ERROR("Error in epoll_wait! %s", strerror(errno));
+        err = errno;
+        return -1;
+      }
+    } else if (fd_cnt == 0) {
+      return 0;
```
Comment (dirk-thomas) [inline clients/roscpp/src/libros/io.cpp:399]: The existing code sets `err` to `0` on success in this case. Should this branch do the same?

**Review by dirk-thomas — COMMENTED** [review]
Beside the inline comments can you please comment on the amount of testing you have done with this patch.

**Comment by astere-cpr** [comment]
> Beside the inline comments can you please comment on the amount of testing you have done with this patch.

The patch has been in continuous use since I submitted the PR on our large fleet testing cluster which runs 24/7. The busiest ROS node on that cluster has around 4K open sockets at any given time.

File: clients/roscpp/src/libros/io.cpp (hunk 2 above)
Comment (astere-cpr) [reply clients/roscpp/src/libros/io.cpp:399]: Done

File: clients/roscpp/src/libros/io.cpp (hunk 1 above)
Comment (astere-cpr) [reply clients/roscpp/src/libros/io.cpp:405]: I changed it to assign the error on the socket being polled, if any

**Comment by dirk-thomas** [comment]
Thanks for the patch and the quick update.

**Pull request merged**